### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by arshiba**
Supply mule receives a new player station builder mode. In this mode, filling the building storage with materials is smoother. To activate the mode, you need to choose allow the build storage only. Not used to supply NPC stations.

### CODE REVIEW

File: aiscripts/mule.lib.evaluate_tradeoffers.xml [hunk 1]
```diff
@@ -109,7 +115,7 @@
 					</do_elseif>
 
 					<!-- how much does the need want? -->
-					<get_ware_reservation object="$buyoffer.owner" type="buy" ware="$buyoffer.ware" result="$needReservations" />
+					<!--<get_ware_reservation object="$buyoffer.owner" type="buy" ware="$buyoffer.ware" result="$needReservations" />-->
```
Comment (bhayden53) [inline aiscripts/mule.lib.evaluate_tradeoffers.xml:118]: Why is this commented out?

File: aiscripts/mule.lib.evaluate_tradeoffers.xml [hunk 2]
```diff
@@ -130,13 +136,24 @@
...
-					<!-- how much profit for this trade? -->
-					<!-- If we use our own stations, we adjust the supply price by the sameFactionBuyMod to allow preferential treatment of player/owned supply chains -->
-					<do_if value="$selloffer.seller.owner == $shipEntity.owner">
-						<set_value name="$currentProfit" exact="$cargoHauled * ($buyoffer.unitprice - $selloffer.unitprice*$sameFactionBuyMod/100.0)" />
+					<do_if value="$dedicatedPlayerBuilder" >
+						<set_value name="$targetAmount" exact="$buyoffer.buyer.cargo.{$buyoffer.ware}.target" />
+						<!--calculate fill level region: 0%..33% = 3; 33%..67% = 2; 67%..100% = 1 -->
+						<!--REMIND (don't optimize) desired amount / target amount equals to the stock level when the construction is fully funded only-->
+						<set_value name="$fillLevelRegion" exact="((($buyoffer.desiredamount)f / ([$targetAmount,1].max)f - 0.0000001) * 3.0 + 1.0)i" />
+						<set_value name="$currentProfit" exact="$fillLevelRegion * 1Cr" />
```
Comment (bhayden53) [inline aiscripts/mule.lib.evaluate_tradeoffers.xml:144]: So the wares that are most needed have a profit of 3Cr, and the least needed a profit of 1Cr? 

I don't think this is much different from just letting the game manage the prices based on storage levels, is it?

**Comment by bhayden53** [comment]
@arshiba Everything looks fine to me, but can you go into a little more detail about the problem you were trying to solve? What you were seeing before and how your changes improve it?

File: aiscripts/mule.lib.evaluate_tradeoffers.xml (hunk 1 above)
Comment (arshiba) [reply aiscripts/mule.lib.evaluate_tradeoffers.xml:118]: Result is not used => CPU time waste. Or maybe i don't know about a side effect of this call.

Comment (bhayden53) [reply aiscripts/mule.lib.evaluate_tradeoffers.xml:118]: Ah, right. I had problems in the past with Mules picking up more of a ware than the destination would buy, and then the trade would error out into the debug log and bad things would happen. I went through a couple of attempts trying to solve that, but I think in the end, the call to clamp_trade_amount right after this is probably what is taking care of that.

File: aiscripts/mule.lib.evaluate_tradeoffers.xml (hunk 2 above)
Comment (arshiba) [reply aiscripts/mule.lib.evaluate_tradeoffers.xml:144]: Profit in this case is a priority. Most needed will be supplied first. Plus a randomizer to fill the storage in equal quantities. Auto price management doesn't work when you are actively drain all the universe by building huge stations.

**Comment by arshiba** [comment]
Problem is:
step 1 plan a huge station;
step 2 assign a fleet to supply build storage;
step 3 wait for half a hour;
step 4 construction doesn't even start.
WTF?
I look at the storage and see something like:
	advanced electronics 3k/7k
	claytronics 5k/25k
	field coils 3k/10k
	hull parts 0/100k (!!!)
	... etc
All because there are a plenty of electronics and coils and very little of hull parts and turret components in NPC stations. In this case the first ones will make more profit than the second ones all the time. It's not good.

**Comment by bhayden53** [comment]
Ah, yeah, I see. By prioritizing the lowest supply it ensures that constructions is always able to continue instead of getting stuck at a module or something. That seems good to me. Thanks for the contribution!

**Pull request merged**